File: GameCreator.Framework.Gml/Misc/NodeVisitor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace GameCreator.Framework.Gml.Misc
{
    public abstract class NodeVisitor
    {
        #region Main Traversal
        public void VisitExpression(Expression e)
        {
            switch (e.Kind)
            {
                case ExpressionKind.Addition:
                case ExpressionKind.BitwiseAnd:
                case ExpressionKind.BitwiseOr:
                case ExpressionKind.BitwiseXor:
                case ExpressionKind.Div:
                case ExpressionKind.Divide:
                case ExpressionKind.Equality:
                case ExpressionKind.GreaterThan:
                case ExpressionKind.GreaterThanOrEqual:
                case ExpressionKind.LessThan:
                case ExpressionKind.LessThanOrEqual:
                case ExpressionKind.LogicalAnd:
                case ExpressionKind.LogicalOr:
                case ExpressionKind.LogicalXor:
                case ExpressionKind.Mod:
                case ExpressionKind.Multiply:
                case ExpressionKind.NotEqual:
                case ExpressionKind.ShiftLeft:
                case ExpressionKind.ShiftRight:
                case ExpressionKind.Subtraction:
                    VisitBinaryExpression(e as BinaryExpression);
                    break;
                case ExpressionKind.Access:
                    VisitAccess(e as Access);
                    break;
                case ExpressionKind.Call:
                    VisitCall(e as Call);
                    break;
                case ExpressionKind.Complement:
                case ExpressionKind.Minus:
                case ExpressionKind.Not:
                case ExpressionKind.Plus:
                    VisitUnary(e as UnaryExpression);
                    break;
                case ExpressionKind.Constant:
                    VisitConstant(e as Constant);
                    break;
                case ExpressionKind.Grouping:
                    VisitExpression((e as Grouping).InnerExpression);
                    break;
                case ExpressionKind.None:
                    VisitEmptyExpression(e as Expression);
                    break;
                case ExpressionKind.Id:
                    throw new InvalidOperationException();
            }
        }

        public void VisitStatement(Statement s)
        {
            switch (s.Kind)
            {
                case StatementKind.AdditionAssignment:
                case StatementKind.AndAssignment:
                case StatementKind.Assignment:
                case StatementKind.DivideAssignment:
                case StatementKind.MultiplyAssignment:
                case StatementKind.OrAssignment:
                case StatementKind.SubtractionAssignment:
                case StatementKind.XorAssignment:
                    VisitAssignment(s as Assignment);
                    break;
                case StatementKind.Break:
                    VisitBreak(s as Break);
                    break;
                case StatementKind.Call:
                    VisitCallStatement(s as CallStatement);
                    break;
                case StatementKind.Case:
                    VisitCaseLabel(s as Case);
                    break;
                case StatementKind.Continue:
                    VisitContinue(s as Continue);
                    break;
                case StatementKind.Default:
                    VisitDefaultCaseLabel(s as Default);
                    break;
                case StatementKind.Do:
                    VisitDo(s as Do);
                    break;
                case StatementKind.Exit:
                    VisitExit(s as Exit);
                    break;
                case StatementKind.For:
                    VisitFor(s as For);
                    break;
                case StatementKind.Globalvar:
                    VisitGlobalvar(s as Globalvar);
                    break;
                case StatementKind.If:
                    VisitIf(s as If);
                    break;
                case StatementKind.Nop:
                    VisitNop(s as Nop);
                    break;
                case StatementKind.Repeat:
                    VisitRepeat(s as Repeat);
                    break;
                case StatementKind.Return:
                    VisitReturn(s as Return);
                    break;
                case StatementKind.Sequence:
                    VisitSequence(s as Sequence);
                    break;
                case StatementKind.Switch:
                    VisitSwitch(s as Switch);
                    break;
                case StatementKind.Var:
                    VisitVar(s as Var);
                    break;
                case StatementKind.While:
                    VisitWhile(s as While);
                    break;
                case StatementKind.With:
                    VisitWith(s as With);
                    break;
            }
        }
        #endregion

        #region Abstract Expression Visitors
        public abstract void VisitBinaryExpression(BinaryExpression binaryExpression);

        public abstract void VisitConstant(Constant constant);

        public abstract void VisitUnary(UnaryExpression unaryExpression);

        public abstract void VisitCall(Call call);

        public abstract void VisitAccess(Access access);

        public abstract void VisitEmptyExpression(Expression expression);
        #endregion

        #region Abstract Statement Visitors
        public abstract void VisitWith(With with);

        public abstract void VisitWhile(While p);

        public abstract void VisitVar(Var var);

        public abstract void VisitSwitch(Switch p);

        public abstract void VisitSequence(Sequence sequence);

        public abstract void VisitReturn(Return p);

        public abstract void VisitRepeat(Repeat repeat);

        public abstract void VisitNop(Nop nop);

        public abstract void VisitIf(If p);

        public abstract void VisitGlobalvar(Globalvar globalvar);

        public abstract void VisitFor(For p);

        public abstract void VisitExit(Exit exit);

        public abstract void VisitDo(Do p);

        public abstract void VisitDefaultCaseLabel(Default p);

        public abstract void VisitContinue(Continue p);

        public abstract void VisitCaseLabel(Case p);

        public abstract void VisitCallStatement(CallStatement callStatement);

        public abstract void VisitBreak(Break p);

        public abstract void VisitAssignment(Assignment assignment);
        #endregion
    }
}

File: JS/NodeVisitor.js
var enums = require('./ParseEnums.js');
var parse = require('./Parser.js');

exports.NodeVisitor = function () {
    var self = this;

    var ExpressionKind = enums.ExpressionKind,
        StatementKind = enums.StatementKind,
        Access = parse.Access,
        Addition = parse.Addition,
        BitwiseAnd = parse.BitwiseAnd,
        BitwiseOr = parse.BitwiseOr,
        BitwiseXor = parse.BitwiseXor,
        Call = parse.Call,
        Complement = parse.Complement,
        Constant = parse.Constant,
        Div = parse.Div,
        Divide = parse.Divide,
        Equality = parse.Equality,
        GreaterThan = parse.GreaterThan,
        GreaterThanOrEqual = parse.GreaterThanOrEqual,
        Grouping = parse.Grouping,
        Id = parse.Id,
        LessThan = parse.LessThan,
        LessThanOrEqual = parse.LessThanOrEqual,
        LogicalAnd = parse.LogicalAnd,
        LogicOr = parse.LogicOr,
        LogicalXor = parse.LogicalXor,
        Minus = parse.Minus,
        Mod = parse.Mod,
        Multiply = parse.Multiply,
        Not = parse.Not,
        NotEqual = parse.NotEqual,
        Plus = parse.Plus,
        ShiftLeft = parse.ShiftLeft,
        ShiftRight = parse.ShiftRight,
        Subtraction = parse.Subtraction,
        AdditionAssignment = parse.AdditionAssignment,
        AndAssignment = parse.AndAssignment,
        Assignment = parse.Assignment,
        Break = parse.Break,
        CallStatement = parse.CallStatement,
        Case = parse.Case,
        Continue = parse.Continue,
        Default = parse.Default,
        DivideAssignment = parse.DivideAssignment,
        Do = parse.Do,
        Exit = parse.Exit,
        For = parse.For,
        Globalvar = parse.Globalvar,
        If = parse.If,
        MultiplyAssignment = parse.MultiplyAssignment,
        Nop = parse.Nop,
        OrAssignment = parse.OrAssignment,
        Repeat = parse.Repeat,
        Return = parse.Return,
        Sequence = parse.Sequence,
        SubtractionAssignment = parse.SubtractionAssignment,
        Switch = parse.Switch,
        Var = parse.Var,
        While = parse.While,
        With = parse.With,
        Statement = parse.Statement;

    // Main Traversal
    this.VisitNode = function(node) {
        if (node instanceof Statement)
            self.VisitStatement(node);
        else
            self.VisitExpression(node);
    }

    this.VisitExpression = function(e) {
        switch (e.Kind)
        {
            case ExpressionKind.Addition:
            case ExpressionKind.BitwiseAnd:
            case ExpressionKind.BitwiseOr:
            case ExpressionKind.BitwiseXor:
            case ExpressionKind.Div:
            case ExpressionKind.Divide:
            case ExpressionKind.Equality:
            case ExpressionKind.GreaterThan:
            case ExpressionKind.GreaterThanOrEqual:
            case ExpressionKind.LessThan:
            case ExpressionKind.LessThanOrEqual:
            case ExpressionKind.LogicalAnd:
            case ExpressionKind.LogicalOr:
            case ExpressionKind.LogicalXor:
            case ExpressionKind.Mod:
            case ExpressionKind.Multiply:
            case ExpressionKind.NotEqual:
            case ExpressionKind.ShiftLeft:
            case ExpressionKind.ShiftRight:
            case ExpressionKind.Subtraction:
                self.VisitBinaryExpression(e);
                break;
            case ExpressionKind.Complement:
            case ExpressionKind.Minus:
            case ExpressionKind.Not:
            case ExpressionKind.Plus:
                self.VisitUnary(e);
                break;
            case ExpressionKind.Access:
                self.VisitAccess(e);
                break;
            case ExpressionKind.Call:
                self.VisitCall(e);
                break;
            case ExpressionKind.Constant:
                self.VisitConstant(e);
                break;
            case ExpressionKind.Grouping:
                self.VisitGrouping(e);
                break;
            case ExpressionKind.None:
            case ExpressionKind.Id:
                throw new Error();
        }
    }

    this.VisitStatement = function(s) {
        switch (s.Kind)
        {
            case StatementKind.AdditionAssignment:
            case StatementKind.AndAssignment:
            case StatementKind.Assignment:
            case StatementKind.DivideAssignment:
            case StatementKind.MultiplyAssignment:
            case StatementKind.OrAssignment:
            case StatementKind.SubtractionAssignment:
            case StatementKind.XorAssignment:
                self.VisitAssignment(s);
                break;
            case StatementKind.Break:
                self.VisitBreak(s);
                break;
            case StatementKind.Call:
                self.VisitCallStatement(s);
                break;
            case StatementKind.Case:
                self.VisitCaseLabel(s);
                break;
            case StatementKind.Continue:
                self.VisitContinue(s);
                break;
            case StatementKind.Default:
                self.VisitDefaultCaseLabel(s);
                break;
            case StatementKind.Do:
                self.VisitDo(s);
                break;
            case StatementKind.Exit:
                self.VisitExit(s);
                break;
            case StatementKind.For:
                self.VisitFor(s);
                break;
            case StatementKind.Globalvar:
                self.VisitGlobalvar(s);
                break;
            case StatementKind.If:
                self.VisitIf(s);
                break;
            case StatementKind.Nop:
                self.VisitNop(s);
                break;
            case StatementKind.Repeat:
                self.VisitRepeat(s);
                break;
            case StatementKind.Return:
                self.VisitReturn(s);
                break;
            case StatementKind.Sequence:
                self.VisitSequence(s);
                break;
            case StatementKind.Switch:
                self.VisitSwitch(s);
                break;
            case StatementKind.Var:
                self.VisitVar(s);
                break;
            case StatementKind.While:
                self.VisitWhile(s);
                break;
            case StatementKind.With:
                self.VisitWith(s);
                break;
        }
    }

    /*
     * The following are abstract functions. Copy-paste them to use them in your implementation.
     */

    // Abstract expression visitors

    this.VisitBinaryExpression = function(binaryExpression) {

    };

    this.VisitConstant = function(constant) {

    };

    this.VisitUnary = function(unaryExpression) {

    };

    this.VisitCall = function(call) {

    };

    this.VisitAccess = function(access) {

    };

    this.VisitGrouping = function(expression) {

    };

    // Abstract Statement Visitors
    this.VisitWith = function(_with) {

    };

    this.VisitWhile = function(p) {

    };

    this.VisitVar = function(_var) {

    };

    this.VisitSwitch = function(p) {

    };

    this.VisitSequence = function(sequence) {
        self.VisitNode(sequence.First);
        self.VisitNode(sequence.Second);
    };

    this.VisitReturn = function(p) {

    };

    this.VisitRepeat = function(repeat) {

    };

    this.VisitNop = function(nop) {

    };

    this.VisitIf = function(p) {

    };

    this.VisitGlobalvar = function(globalvar) {

    };

    this.VisitFor = function(p) {

    };

    this.VisitExit = function(exit) {

    };

    this.VisitDo = function(p) {

    };

    this.VisitDefaultCaseLabel = function(p) {

    };

    this.VisitContinue = function(p) {

    };

    this.VisitCaseLabel = function(p) {

    };

    this.VisitCallStatement = function(callStatement) {

    };

    this.VisitBreak = function(p) {

    };

    this.VisitAssignment = function(assignment) {

    };
};
